collections/schema.go

```diff
@@ -4,8 +4,64 @@ import (
 	"fmt"
 	storetypes "github.com/cosmos/cosmos-sdk/store/types"
 	"regexp"
+	"strings"
+
+	"github.com/hashicorp/go-multierror"
 )
 
+// SchemaBuilder is used for building schemas. The Build method should always
+// be called after all collections have been initialized. Initializing new
+// collections with the builder after initialization will result in panics.
+type SchemaBuilder struct {
+	schema *Schema
+	err    *multierror.Error
+}
+
+// NewSchemaBuilder creates a new schema builder from the provided store key.
+// Callers should always call the SchemaBuilder.Build method when they are
+// done adding collections to the schema.
+func NewSchemaBuilder(storeKey storetypes.StoreKey) *SchemaBuilder {
+	return &SchemaBuilder{
+		schema: &Schema{
+			storeKey:            storeKey,
+			collectionsByName:   map[string]collection{},
+			collectionsByPrefix: map[string]collection{},
+		},
+	}
+}
+
+// Build should be called after all collections that are part of the schema
+// have been initialized in order to get a reference to the Schema. It is
+// important to check the returned error for any initialization errors.
+// The SchemaBuilder CANNOT be used after Build is called - doing so will
+// result in panics.
+func (s *SchemaBuilder) Build() (Schema, error) {
+	if s.err != nil {
+		return Schema{}, s.err
+	}
+
+	// check for any overlapping prefixes
+	for prefix := range s.schema.collectionsByPrefix {
+		for prefix2 := range s.schema.collectionsByPrefix {
+			// don't compare the prefix to itself
+			if prefix == prefix2 {
+				continue
+			}
+
+			// if one prefix is the prefix of the other we have an overlap and
+			// this schema is corrupt
+			if strings.HasPrefix(prefix, prefix2) {
+				return Schema{}, fmt.Errorf("schema has overlapping prefixes 0x%x and 0x%x", prefix, prefix2)
+			}
+		}
+	}
+
+	schema := *s.schema
+	s.schema = nil // this makes the builder unusable
```

Comment:
maybe an explicit panic check is better to debug than a nil pointer dereference I think

Reply:
added, can you check?